wiring/inc/spark_wiring_variant.h

```diff
@@ -0,0 +1,1058 @@
+/*
+ * Copyright (c) 2023 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <variant>
+#include <type_traits>
+#include <algorithm>
+#include <charconv>
+#include <cstdint>
+
+#include "spark_wiring_string.h"
+#include "spark_wiring_vector.h"
+#include "spark_wiring_map.h"
+
+#include "debug.h"
+
+class Stream;
+class Print;
+
+namespace spark {
+
+class JSONValue;
+
+} // namespace spark
+
+namespace particle {
+
+using spark::JSONValue;
+
+class Variant;
+
+/**
+ * An array of `Variant` values.
+ */
+typedef Vector<Variant> VariantArray;
+
+/**
+ * A map of named `Variant` values.
+ */
+typedef Map<String, Variant> VariantMap;
+
+namespace detail {
+
+template<typename T>
+struct IsComparableWithVariant {
+    // TODO: This is not ideal as we'd like Variant to be comparable with any type as long as it's
+    // comparable with one of the Variant's alternative types
+    static constexpr bool value = std::is_same_v<T, std::monostate> || std::is_arithmetic_v<T> || std::is_same_v<T, const char*> ||
+            std::is_same_v<T, String> || std::is_same_v<T, VariantArray> || std::is_same_v<T, VariantMap>;
+};
+
+} // namespace detail
+
+/**
+ * A class acting as a union for certain common data types.
+ */
+class Variant {
+public:
+    /**
+     * Supported alternative types.
+     */
+    enum Type {
+        NULL_, ///< Null type (`std::monostate`).
+        BOOL, ///< `bool`.
+        INT, ///< `int`.
+        UINT, ///< `unsigned`.
+        INT64, ///< `int64_t`.
+        UINT64, ///< `uint64_t`.
+        DOUBLE, ///< `double`.
```

Comment:
All `float` types will have to be promoted?

Reply:
Yes. This class isn't very efficient for storing values of primitive types in the first place so it doesn't make much sense to support all possible variations of those types. 32- and 64-bit integers are handled separately because 32-bit arithmetic is handled in hardware which is not the case for `float` and `double`.